compiler/src/iree/compiler/Dialect/Flow/Transforms/FormDispatchWorkgroups.cpp

```diff
@@ -0,0 +1,371 @@
+  SmallVector<Flow::DispatchWorkgroupsOp> result;
+  for (auto regionOp : regionOps) {
+    if (failed(cloneProducers(rewriter, regionOp))) return failure();
+    auto maybeWorkgroupOp =
+        Flow::rewriteFlowDispatchRegionToFlowDispatchWorkgroups(regionOp,
+                                                                rewriter);
+    if (failed(maybeWorkgroupOp)) return failure();
+    result.push_back(*maybeWorkgroupOp);
+  }
+  return result;
+}
+
+/// Return `true` if the given op is contained in DispatchWorkgroupsOp or in a
+/// DispatchRegionOp.
+static bool isInDispatchRegion(Operation *op) {
+  return op->getParentOfType<Flow::DispatchWorkgroupsOp>() ||
+         op->getParentOfType<Flow::DispatchRegionOp>();
+}
+
+/// Wrap a single op in a DispatchWorkgroupsOp.
+static FailureOr<Flow::DispatchWorkgroupsOp> wrapInWorkgroupsOp(
+    mlir::TensorDimTrackingRewriter &rewriter, Operation *op,
+    bool generateWorkloadRegion) {
+  // Compute workload.
+  Optional<Flow::WorkloadBuilder> workloadBuilder = llvm::None;
+  if (generateWorkloadRegion) {
+    auto maybeBuilder =
+        iree_compiler::IREE::Flow::getWorkloadBuilder(rewriter, op);
+    if (failed(maybeBuilder)) return failure();
+    workloadBuilder = *maybeBuilder;
+  }
+
+  // Simplify tensor::DimOps.
+  SmallVector<tensor::DimOp> dimOps = rewriter.getTensorDimOps();
+  if (failed(iree_compiler::IREE::Flow::simplifyDimOps(
+          rewriter, rewriter.getTensorDimOps())))
+    return failure();
+
+  // Wrap operation.
+  auto regionOp = Flow::wrapOpInDispatchRegion(rewriter, op, workloadBuilder);
+  if (failed(regionOp)) return failure();
+  if (failed(cloneProducers(rewriter, *regionOp))) return failure();
+  auto workgroupsOp = Flow::rewriteFlowDispatchRegionToFlowDispatchWorkgroups(
+      *regionOp, rewriter);
+  if (failed(workgroupsOp)) return failure();
+  return *workgroupsOp;
+}
+
+/// Wrap all given ops in a DispatchWorkgroupsOp.
+static FailureOr<SmallVector<Flow::DispatchWorkgroupsOp>> wrapInWorkgroupsOp(
+    mlir::TensorDimTrackingRewriter &rewriter, SmallVector<Operation *> rootOps,
+    bool generateWorkloadRegion) {
+  SmallVector<Flow::DispatchWorkgroupsOp> result;
+  for (Operation *rootOp : rootOps) {
+    auto workgroupsOp =
+        wrapInWorkgroupsOp(rewriter, rootOp, generateWorkloadRegion);
+    if (failed(workgroupsOp)) return failure();
+    result.push_back(*workgroupsOp);
+  }
+  return result;
+}
+
+/// Wrap all ops of the given types that are direct children of the given op
+/// in DispatchWorkgroupsOps.
+template <typename... OpTys>
+static FailureOr<SmallVector<Flow::DispatchWorkgroupsOp>> wrapInWorkgroupsOp(
+    mlir::TensorDimTrackingRewriter &rewriter, Operation *op,
+    bool generateWorkloadRegion) {
+  // Find ops of type OpTys.
+  SmallVector<Operation *> rootOps;
+  for (Region &r : op->getRegions())
+    for (Block &b : r.getBlocks())
+      for (Operation &op : b)
+        if (isa<OpTys...>(&op)) rootOps.push_back(&op);
```

Comment:
Does `Region::getOps()` work for this case? That could help less nested loops. E.g.,

```
for (Region &r : op->getRegions()) {
  for (auto &op: r->getOps()) {
    if (isa<OpTys..
  }
}
```

also, we have to wrap braces for this case.